src/analysis/calculus/lagrange_multipliers.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2021 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Yury Kudryashov
+-/
+import analysis.calculus.local_extr
+import analysis.calculus.implicit
+
+/-!
+# Lagrange multipliers
+
+In this file we formalize the
+[Lagrange multipliers](https://en.wikipedia.org/wiki/Lagrange_multiplier) method of solving
+conditional extremum problems: if a function `φ` has a local extremum at `x₀` on the set
+`f ⁻¹' {f x₀}`, `f x = (f₀ x, ..., fₙ₋₁ x)`, then the differentials of `fₖ` and `φ` are linearly
+dependent. First we formulate a geometric version of this theorem which does not rely on the
+target space being `ℝⁿ`, then restate it in terms of coordinates.
+
+## TODO
+
+Formalize Karush-Kuhn-Tucker theorem
+
+## Tags
+
+lagrange multiplier, local extremum
+
+-/
+
+open filter set
+open_locale topological_space filter big_operators
+variables {E F : Type*} [normed_group E] [normed_space ℝ E] [complete_space E]
+  [normed_group F] [normed_space ℝ F] [complete_space F]
+  {f : E → F} {φ : E → ℝ} {x₀ : E} {f' : E →L[ℝ] F} {φ' : E →L[ℝ] ℝ}
+
+/-- Lagrange multipliers theorem: if `φ : E → ℝ` has a local extremum on the set `{x | f x = f x₀}`
+at `x₀`, both `f : E → F` and `φ` are strictly differentiable at `x₀`, and the codomain of `f` is
+a complete space, then the linear map `x ↦ (f' x, φ' x)` is not surjective. -/
+lemma is_local_extr_on.range_ne_top_of_has_strict_fderiv_at
+  (hextr : is_local_extr_on φ {x | f x = f x₀} x₀) (hf' : has_strict_fderiv_at f f' x₀)
```

Comment:
Isn't the theorem true if you only assume that it has a derivative (I mean, not a strict one)? You don't need local surjectivity, you just need to find a direction where you can increase a little bit the function, so linear approximation should be enough for this. (Disclaimer: I have not tried to write a precise proof of this :-)

Reply:
I just found a counterexample on the internet (https://www.jstor.org/stable/2133110, Example 1) so discard my previous comment: you need something more than mere differentiability.

Reply:
As far as I understand, this is not the strongest possible statement but I wanted to formalize some form of Lagrange multipliers before we have enough theory (e.g., hyperplane separation theorem) to easily formalize some stronger version.